cloud/grpc.go

```diff
@@ -0,0 +1,77 @@
+package cloud
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/pkg/errors"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+)
+
+func (c *client) withAuth(ctx context.Context) context.Context {
+	return metadata.AppendToOutgoingContext(ctx, "authorization", fmt.Sprintf("Bearer %s", c.authToken))
+}
+
+func (c *client) UnaryAuthInterceptor() grpc.UnaryClientInterceptor {
+	return func(ctx context.Context, method string, req, reply interface{}, cc *grpc.ClientConn, invoker grpc.UnaryInvoker, opts ...grpc.CallOption) error {
+		if time.Now().UTC().After(c.authTokenExpiry) {
+			err := c.Authenticate(ctx)
+			if err != nil {
+				return errors.Wrap(err, "failed refreshing expired token")
+			}
+		}
+		err := invoker(ctx, method, req, reply, cc, opts...)
+		if err != nil {
+			s, ok := status.FromError(err)
+			if ok && s.Code() == codes.Unauthenticated {
```

Comment:
Is it possible to bake some of this auth retry logic into the gRPC retrier code?

Reply:
I didn't see a clean way. Would be cool if the library supported hooking-in custom functions to run before retrying (and also modify the retry limits from within that func).